RunDataFrameAnalyticsIT failures

There are a few tests that fail: 

org.elasticsearch.xpack.ml.integration.RunDataFrameAnalyticsIT testOutlierDetectionWithEnoughDocumentsToScroll
org.elasticsearch.xpack.ml.integration.RunDataFrameAnalyticsIT testOutlierDetectionWithFewDocuments
org.elasticsearch.xpack.ml.integration.RunDataFrameAnalyticsIT testOutlierDetectionWithMultipleSourceIndices
org.elasticsearch.xpack.ml.integration.RunDataFrameAnalyticsIT testOutlierDetectionWithPreExistingDestIndex

with the same error. 

<!--
Please fill out the following information, and ensure you have attempted
to reproduce locally
-->

**Build scan**:
[7.10.0] https://gradle-enterprise.elastic.co/s/dp2u6q52kyxus
[7.9.1] https://gradle-enterprise.elastic.co/s/dlgajwjbzcmli
[7.10.0] https://gradle-enterprise.elastic.co/s/w57smywps3wlq
[7.9.1] https://gradle-enterprise.elastic.co/s/at3whtu5np6r6
https://gradle-enterprise.elastic.co/s/c5egkz647juis
 
and

[7.9.1] https://gradle-enterprise.elastic.co/s/jbaq7v3utks6q
[7.10.0] https://gradle-enterprise.elastic.co/s/bnbd27r4httos
[7.10.0] https://gradle-enterprise.elastic.co/s/si6k6omt5gjfe
[7.9.1] https://gradle-enterprise.elastic.co/s/blgn5kzzmp7pg

[7.10.0] https://gradle-enterprise.elastic.co/s/d5w6ctq3li7va
https://gradle-enterprise.elastic.co/s/3o4d53arbbfia

**Repro line**:

Example: 
```
REPRODUCE WITH: gradlew ':x-pack:plugin:ml:qa:native-multi-node-tests:integTest' --tests "org.elasticsearch.xpack.ml.integration.RunDataFrameAnalyticsIT.testOutlierDetectionWithFewDocuments" \
  -Dtests.seed=E6859593F0B2A900 \
  -Dtests.security.manager=true \
  -Dtests.locale=es-EC \
  -Dtests.timezone=Canada/Newfoundland \
  -Druntime.java=11
```

**Reproduces locally?**:
Didn't try ( windows)

**Applicable branches**:
master

**Failure history**:

FAiled twice today but hasnt been failing recently with this assertion error. 

**Failure excerpt**:

```

org.elasticsearch.xpack.ml.integration.RunDataFrameAnalyticsIT > testOutlierDetectionWithFewDocuments FAILED
    java.lang.AssertionError: progress is complete: [{"phase":"reindexing","progress_percent":1}, {"phase":"loading_data","progress_percent":100}, {"phase":"computing_outliers","progress_percent":100}, {"phase":"writing_results","progress_percent":100}]
    Expected: is <true>
         but: was <false>
        at __randomizedtesting.SeedInfo.seed([E6859593F0B2A900:58D5761152F12CEA]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.elasticsearch.xpack.ml.integration.MlNativeDataFrameAnalyticsIntegTestCase.assertProgressComplete(MlNativeDataFrameAnalyticsIntegTestCase.java:227)
        at org.elasticsearch.xpack.ml.integration.RunDataFrameAnalyticsIT.testOutlierDetectionWithFewDocuments(RunDataFrameAnalyticsIT.java:143)
```

--- a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetDataFrameAnalyticsStatsAction.java
+++ b/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetDataFrameAnalyticsStatsAction.java
@@ -119,6 +119,7 @@ public class TransportGetDataFrameAnalyticsStatsAction
             }, listener::onFailure
         );
 
+        // We must update the progress of the reindexing task as it might be stale
         task.updateReindexTaskProgress(reindexingProgressListener);
     }
 

--- a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsTask.java
+++ b/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsTask.java
@@ -290,7 +290,7 @@ public class DataFrameAnalyticsTask extends AllocatedPersistentTask implements S
 
         String progressDocId = StoredProgress.documentId(jobId);
 
-        // Step 3: Run the runnable provided as the argument
+        // Step 4: Run the runnable provided as the argument
         ActionListener<IndexResponse> indexProgressDocListener = ActionListener.wrap(
             indexResponse -> {
                 LOGGER.debug("[{}] Successfully indexed progress document", jobId);
@@ -303,7 +303,7 @@ public class DataFrameAnalyticsTask extends AllocatedPersistentTask implements S
             }
         );
 
-        // Step 2: Create or update the progress document:
+        // Step 3: Create or update the progress document:
         //   - if the document did not exist, create the new one in the current write index
         //   - if the document did exist, update it in the index where it resides (not necessarily the current write index)
         ActionListener<SearchResponse> searchFormerProgressDocListener = ActionListener.wrap(
@@ -331,14 +331,26 @@ public class DataFrameAnalyticsTask extends AllocatedPersistentTask implements S
             }
         );
 
-        // Step 1: Search for existing progress document in .ml-state*
-        SearchRequest searchRequest =
-            new SearchRequest(AnomalyDetectorsIndex.jobStateIndexPattern())
-                .source(
-                    new SearchSourceBuilder()
-                        .size(1)
-                        .query(new IdsQueryBuilder().addIds(progressDocId)));
-        executeAsyncWithOrigin(client, ML_ORIGIN, SearchAction.INSTANCE, searchRequest, searchFormerProgressDocListener);
+        // Step 2: Search for existing progress document in .ml-state*
+        ActionListener<Void> reindexProgressUpdateListener = ActionListener.wrap(
+            aVoid -> {
+                SearchRequest searchRequest =
+                    new SearchRequest(AnomalyDetectorsIndex.jobStateIndexPattern())
+                        .source(
+                            new SearchSourceBuilder()
+                                .size(1)
+                                .query(new IdsQueryBuilder().addIds(progressDocId)));
+                executeAsyncWithOrigin(client, ML_ORIGIN, SearchAction.INSTANCE, searchRequest, searchFormerProgressDocListener);
+            },
+            e -> {
+                LOGGER.error(new ParameterizedMessage(
+                    "[{}] cannot persist progress as an error occurred while updating reindexing task progress", taskParams.getId()), e);
+                runnable.run();
+            }
+        );
+
+        // Step 1: Update reindexing progress as it could be stale
+        updateReindexTaskProgress(reindexProgressUpdateListener);
     }
 
     /**

--- a/x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsTaskTests.java
+++ b/x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsTaskTests.java
@@ -46,6 +46,7 @@ import java.util.List;
 import java.util.Map;
 
 import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasSize;
 import static org.mockito.Matchers.any;
 import static org.mockito.Matchers.eq;
 import static org.mockito.Matchers.same;
@@ -216,8 +217,9 @@ public class DataFrameAnalyticsTaskTests extends ESTestCase {
         PersistentTasksService persistentTasksService = new PersistentTasksService(clusterService, mock(ThreadPool.class), client);
         TaskManager taskManager = mock(TaskManager.class);
 
+        // We leave reindexing progress here to zero in order to check it is updated before it is persisted
         List<PhaseProgress> progress = List.of(
-            new PhaseProgress(ProgressTracker.REINDEXING, 100),
+            new PhaseProgress(ProgressTracker.REINDEXING, 0),
             new PhaseProgress(ProgressTracker.LOADING_DATA, 100),
             new PhaseProgress(ProgressTracker.WRITING_RESULTS, 30));
 
@@ -239,6 +241,7 @@ public class DataFrameAnalyticsTaskTests extends ESTestCase {
             new DataFrameAnalyticsTask(
                 123, "type", "action", null, Map.of(), client, clusterService, analyticsManager, auditor, taskParams);
         task.init(persistentTasksService, taskManager, "task-id", 42);
+        task.setReindexingFinished();
         Exception exception = new Exception("some exception");
 
         task.setFailed(exception);
@@ -260,7 +263,8 @@ public class DataFrameAnalyticsTaskTests extends ESTestCase {
             try (XContentParser parser = JsonXContent.jsonXContent.createParser(
                 NamedXContentRegistry.EMPTY, DeprecationHandler.IGNORE_DEPRECATIONS, indexRequest.source().utf8ToString())) {
                 StoredProgress parsedProgress = StoredProgress.PARSER.apply(parser, null);
-                assertThat(parsedProgress.get(), equalTo(progress));
+                assertThat(parsedProgress.get(), hasSize(3));
+                assertThat(parsedProgress.get().get(0), equalTo(new PhaseProgress("reindexing", 100)));
             }
 
             verify(client).execute(
@@ -269,7 +273,7 @@ public class DataFrameAnalyticsTaskTests extends ESTestCase {
                     "task-id", 42, new DataFrameAnalyticsTaskState(DataFrameAnalyticsState.FAILED, 42, "some exception"))),
                 any());
         }
-        verifyNoMoreInteractions(client, clusterService, analyticsManager, auditor, taskManager);
+        verifyNoMoreInteractions(client, analyticsManager, auditor, taskManager);
     }
 
     @SuppressWarnings("unchecked")

